src/TorchSharp/Tensor/Tensor.Factories.cs

```diff
@@ -2045,6 +2045,87 @@ public static Tensor tensor(IList<System.Numerics.Complex> rawArray, long dim0,
+        /// </summary>
+        /// <param name="rawArray">The arbitrary array to create the tensor from.</param>
+        /// <param name="dtype">The torch data type.</param>
+        /// <param name="device">The torch device.</param>
+        /// <param name="requiresGrad">Set <value>true</value> if gradients need to be computed for this Tensor; <value>false</value> otherwise.</param>
+        /// <returns>A <see cref="torch.Tensor">torch tensor</see></returns>
+        /// <exception cref="InvalidOperationException">
+        /// When <see cref="Type.GetElementType()">Array.GetType().GetElementType()</see> does not return the .NET element type.
+        /// </exception>
+        /// <exception cref="NotSupportedException">
+        /// When <see cref="Type.GetElementType()">Array.GetType().GetElementType()</see> returns an unsupported .NET element type.
+        /// Supported element types are <see cref="bool" />, <see cref="byte" />, <see cref="sbyte" />, <see cref="short" />,
+        /// <see cref="int" />, <see cref="long" />, <see cref="float" />, <see cref="double" />,
+        /// and <see cref="System.Numerics.Complex" />.
+        /// </exception>
+        /// <example>
+        /// Tensor from array of rank 1
+        /// <code>
+        /// var array = new double[] { { 1, 2, 3, 4, 5, 6, 7, 8 } };
+        /// var tensor = torch.tensor_from_array(rawArray: array, dtype: torch.ScalarType.Float64, device: torch.Device.CPU, requiresGrad: false);
+        /// </code>
+        /// Tensor from array of rank 2
+        /// <code>
+        /// var array = new double[,] { { 1, 2, 3, 4 }, { 5, 6, 7, 8 } };
+        /// var tensor = torch.tensor_from_array(rawArray: array, dtype: torch.ScalarType.Float64, device: torch.Device.CPU, requiresGrad: false);
+        /// </code>
+        /// Tensor from array of rank 3
+        /// <code>
+        /// var array = new double[,,] { { { 1, 2 }, { 3, 4 } }, { { 5, 6 }, { 7, 8 } } };
+        /// var tensor = torch.tensor_from_array(rawArray: array, dtype: torch.ScalarType.Float64, device: torch.Device.CPU, requiresGrad: false);
+        /// </code>
+        /// </example>
+        public static Tensor tensor_from_array(Array rawArray, ScalarType? dtype = null, Device? device = null, bool requiresGrad = false)
+        {
+            // enumerates over all dimensions of the arbitrary array
+            // and returns the length of the dimension
+            IEnumerable<long> GetShape(Array arr)
+            {
+                for (var dim = 0; dim < arr.Rank; dim++) {
+                    var dimLength = arr.GetLength(dim);
+                    yield return dimLength;
+                }
+            }
+            var shape = GetShape(rawArray).ToArray();
+
+            var t = rawArray.GetType().GetElementType();
+            if (t is null) throw new InvalidOperationException($"{nameof(rawArray)}.GetType().GetElementType() returned null.");
+
+            IEnumerable<object> GetElements(Array rawArray)
```

Comment:
This function seem unnecessary. 'Array' already implements IEnumerable, so this just adds overhead, I think.